app/classes/Transvision/Search.php

```diff
@@ -0,0 +1,228 @@
+<?php
+namespace Transvision;
+
+/**
+ * Search class
+ *
+ * Allows searching for data in our repositories using a fluent interface.
+ * Currently, only the regex part (definition of the search) is implemented.
+ * ex:
+ * $search = (new Search)
+ *     ->setSearchTerms('Bookmark this page')
+ *     ->setRegexWholeWords(true)
+ *     ->setRegexCase('sensitive')
+ *     ->setRegexPerfectMatch(false);
+ */
+class Search
+{
+    /**
+     * The trimmed string searched, we keep that one as the canonical reference
+     * @var string
+     */
+    protected $search_terms;
+
+    /**
+     * The generated regex string updated dynamically via updateRegex()
+     * @var string
+     */
+    protected $regex;
+
+    /**
+     * Case sensibility of the regex
+     * @var string
+     */
+    protected $regex_case;
+
+    /**
+     * Consider the space separated string as a single word for search
+     * @var string
+     */
+    protected $regex_whole_words;
+
+    /**
+     * Only return strings that match the search perfectly (case excluded)
+     * @var boolean
+     */
+    protected $regex_perfect_match;
+
+    /**
+     * The search terms for the regex, those differ from $search_terms as
+     * they can be changed dynamically via setRegexSearchTerms()
+     * @var string
+     */
+    protected $regex_search_terms;
+
+    /**
+     * We set the default values for a search
+     */
+    public function __construct()
+    {
+        $this->search_terms = '';
+        $this->regex = '';
+        $this->regex_case = 'i';
+        $this->regex_whole_words = '';
+        $this->regex_perfect_match = false;
+        $this->regex_search_terms = '';
+    }
+
+    /**
+     * Store the searched string in $search_terms and in $regex_search_terms
+     *
+     * @param  string $string String we want to search for
+     * @return $this
+     */
+    public function setSearchTerms($string)
+    {
+        $this->search_terms = trim($string);
+        $this->regex_search_terms = $this->search_terms;
+        $this->updateRegex();
+
+        return $this;
+    }
+
+    /**
+     * Allows setting a new searched term for the regex.
+     * This is mostly useful when you have a multi-words search and need to
+     * loop through all the words to return results.
+     *
+     * @param  string $string The string we want to update the regex for
+     * @return $this
+     */
+    public function setRegexSearchTerms($string)
+    {
+        $this->regex_search_terms = $string;
+        $this->updateRegex();
+
+        return $this;
+    }
+
+    /**
+     * Set the regex case sensibility.
+     *
+     * @param  boolean $flag 'sensitive' == '' in a regex
+     * @return $this
+     */
+    public function setRegexCase($flag)
+    {
+        $this->regex_case = (boolean) $flag ? '' : 'i';
```

Comment:
This is quite confusing: the function is called `setRegexCase`, but it should actually be `setRegexCaseSensitive`.
Proof is that calling `$this->setRegexCase('insensitive')` set the regex as case sensitive.

I would also update the example at the beginning of the class with an explicit true or false.